File: RentalsDownloader/ReportPage.cs
namespace RentalsDownloader
{
	using System.Collections.Generic;
	using System.IO;
	using System.Linq;
	using OfficeOpenXml;
	using WebRobotFramework;

	sealed class ReportPage
	{
		readonly string _title;

		public ReportPage(string title) { _title = title; }

		public List<Rental> Rentals { get; } = new List<Rental>();

		public void Save()
		{
			if (!Rentals.Any())
				return;
			Logger.Log($"Saving report page '{_title}'.");
			LoadExistingSpreadsheetValues(Rentals);
			var rentals =
				Rentals.OrderByDescending(r => r.RatingAsInt)
				       .ThenBy(r => r.AcceptsCats ? 0 : 1)
				       .ThenBy(r => r.RentPerSquareFeet)
				       .ThenByDescending(r => r.SquareFootageAsInt)
				       .ThenByDescending(r => r.BedroomsAsInt)
				       .ThenBy(r => r.HasParking ? 0 : 1)
				       .ThenBy(r => r.RentAsDouble)
				       .ToList();
			var index = 1;
			var rentalsHtml = rentals.Aggregate("", (current, rental) => current + rental.ToHtml(index++));
			File.WriteAllText($"{RentalsDownloaderService.DownloadPath}{_title}.html", Rental.ReportTemplate.Replace("[Rentals]", rentalsHtml));

			SaveSpreadSheet(rentals);
		}

		static string GetWorkSheetValue(WorkSheet worksheet, int row, string columnName, string existingValue)
			=> worksheet.Columns.ContainsKey(columnName) ? worksheet.Worksheet.Cells[row, worksheet.Columns[columnName]].Text : existingValue;

		static void LoadExistingSpreadsheetValues(IEnumerable<Rental> rentals)
		{
			var oldFileInfo = new FileInfo($"{RentalsDownloaderService.DownloadPath}_Favorite Neigboorhoods.xlsx");
			if (!oldFileInfo.Exists)
				return;
			using (var oldExcelSpreadSheet = new ExcelPackage(oldFileInfo))
			{
				var oldWorkSheet = new WorkSheet(oldExcelSpreadSheet.Workbook.Worksheets["Rentals"]);
				var rentalsDictionary = rentals.ToDictionary(r2 => r2.SpreadsheetHyperlink);
				for (var row = 2; row <= oldWorkSheet.Worksheet.Dimension.End.Row; row++)
				{
					var link = oldWorkSheet.Worksheet.Cells[row, oldWorkSheet.Columns["Link"]].Formula;
					Rental rental;
					if (!rentalsDictionary.TryGetValue(link, out rental))
						continue;
					rental.Comments = GetWorkSheetValue(oldWorkSheet, row, "Comments", rental.Comments);
					rental.Rating = GetWorkSheetValue(oldWorkSheet, row, "R", rental.Rating);
					rental.Visited = GetWorkSheetValue(oldWorkSheet, row, "V", rental.Visited);
					rental.Called = GetWorkSheetValue(oldWorkSheet, row, "C", rental.Called);
				}
			}
		}

		void SaveSpreadSheet(IEnumerable<Rental> rentals)
		{
			using (var newExcelSpreadsheet = new ExcelPackage(new FileInfo($"{RentalsDownloaderService.DownloadPath}template\\Rentals.xlsx")))
			{
				var fileInfo = new FileInfo($"{RentalsDownloaderService.DownloadPath}{_title}.xlsx");
				var worksheet = new WorkSheet(newExcelSpreadsheet.Workbook.Worksheets["Rentals"]);
				var rowIndex = 2;
				foreach (var rental in rentals)
					rental.SaveToSpreadsheet(worksheet, rowIndex++);
				newExcelSpreadsheet.SaveAs(fileInfo);
			}
		}
	}
}

File: RentalsDownloader.Service/ReportPage.fs
module ReportPage

open OfficeOpenXml
open System
open System.IO
open WebRobotFramework

let downloadPath = @"C:\Rentals\"

let create title = 
    { title = title
      rentals = list.Empty }

let save page = 
    if page.rentals |> List.isEmpty then ()
    else 
        let loadSpreadSheet rentals = 
            let fileInfo = new FileInfo(String.Format("{0}_Favorite Neighborhoods.xlsx", downloadPath))
            if not fileInfo.Exists then ()
            else 
                use spreadSheet = new ExcelPackage(fileInfo)
                let workSheet = new WorkSheet(spreadSheet.Workbook.Worksheets.["Rentals"])
                let rentalsDictionary = rentals |> toDict (fun r -> r.download.url)
                
                let loadRow row = 
                    let link = workSheet.Worksheet.Cells.[row, workSheet.Columns.["Link"]].Formula
                    let found, rental = rentalsDictionary.TryGetValue(link)
                    if not found then ()
                    else 
                        let getWorksheetValue column existingValue = 
                            if workSheet.Columns.ContainsKey(column) then workSheet.Worksheet.Cells.[row, workSheet.Columns.[column]].Text
                            else existingValue
                        
                        let rentals' = 
                            { rental with comments = getWorksheetValue "Comments" rental.comments
                                          rating = getWorksheetValue "R" rental.rating
                                          visited = getWorksheetValue "V" rental.visited
                                          called = getWorksheetValue "C" rental.called }
                        
                        rentalsDictionary.[link] = rentals' |> ignore
                for row in [ 2..workSheet.Worksheet.Dimension.End.Row ] do
                    loadRow row
        
        let saveSpreadSheet rentals = 
            use spreadSheet = new ExcelPackage(new FileInfo(String.Format(@"{0}template\Rentals.xlsx", downloadPath)))
            let fileInfo = new FileInfo(String.Format("{0}{1}.xlsx", downloadPath, page.title))
            let workSheet = new WorkSheet(spreadSheet.Workbook.Worksheets.["Rentals"])
            let rentalsWithRowIndices = rentals |> assignIndicesFunc (fun i -> i + 2)
            for (i, r) in rentalsWithRowIndices do
                Rental.saveToSpreadsheet workSheet i r
            spreadSheet.SaveAs(fileInfo)
        
        loadSpreadSheet page.rentals
        let rentals = page.rentals |> List.sortBy (fun r -> (-r.ratingInt, boolToInvInt (r.acceptsCats), r.rentPerSqFt, -r.squareFootageInt, -r.bedroomsInt, boolToInvInt r.hasParking, r.rentDouble))
        let rentalHtml = fst (rentals |> List.fold (fun (result, i) rental -> (result + Rental.toHtml i rental, i + 1)) ("", 0))
        File.WriteAllText(String.Format("{0}{1}.html", downloadPath, page.title), Rental.reportTemplate.Replace("[Rentals]", rentalHtml))
        saveSpreadSheet rentals
